blocker/blocker.go

```diff
@@ -198,74 +205,118 @@ func (bl *Blocker) Start() {
+			select {
+			case <-bl.staticCtx.Done():
+				return
+			case <-time.After(retryInterval):
+			}
+			err := bl.RetryFailedSkylinks()
+			if err != nil {
+				bl.staticLogger.Debugf("RetryFailedSkylinks error: %s", err.Error())
+				continue
 			}
-			bl.staticLogger.Warnf("failed to acquire nginx lock")
-			time.Sleep(time.Second)
+			bl.staticLogger.Debugf("RetryFailedSkylinks ran successfully.")
 		}
-		return lockErr
 	}()
-	if err != nil {
-		return errors.AddContext(err, "failed to acquire nginx lock")
-	}
+}
+
+// blockSkylinks blocks the given list of skylinks.
+func (bl *Blocker) blockSkylinks(skylinks []database.BlockedSkylink) (succeeded int, failures int, err error) {
+	batchSize := blockBatchSize
+	start := 0
+
+	// keep track of which skylinks were blocked and which ones failed
+	var blocked []database.BlockedSkylink
+	var failed []database.BlockedSkylink
 
-	// defer a function that releases the lock
+	// defer a function that updates the database and sets return values
 	defer func() {
-		err := os.Remove(bl.staticNginxCachePurgeLockPath)
-		if err != nil {
-			bl.staticLogger.Errorf("failed to release nginx lock, err %v", err)
-		}
+		bErr := bl.staticDB.MarkAsSucceeded(blocked)
+		fErr := bl.staticDB.MarkAsFailed(failed)
+
+		err = errors.Compose(err, bErr, fErr)
+		succeeded = len(blocked)
+		failures = len(failed)
 	}()
 
-	// open the nginx cache list file
-	f, err := os.OpenFile(bl.staticNginxCachePurgerListPath, os.O_RDWR|os.O_CREATE, 0644)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		e1 := f.Sync()
-		e2 := f.Close()
-		if e1 != nil || e2 != nil {
-			bl.staticLogger.Warnf("Failed to sync and close nginx cache purger list: %s", errors.Compose(e1, e2).Error())
+	for start < len(skylinks) {
+		// check whether we need to escape
+		select {
+		case <-bl.staticCtx.Done():
+			return
+		default:
 		}
-	}()
-	for _, s := range sls {
-		_, err = f.WriteString(s + "\n")
+
+		// calculate the end of the batch range
+		end := start + batchSize
+		if end > len(skylinks) {
+			end = len(skylinks)
+		}
+
+		// grab all skylinks for this batch
+		batch := make([]string, end-start)
+		for i, sl := range skylinks[start:end] {
+			batch[i] = sl.Skylink
+		}
+
+		// send the batch to skyd, if an error occurs and the current batch size
+		// is greater than one, we simply retry with a smaller batch size
+		err = bl.staticSkydAPI.BlockSkylinks(batch)
+
+		// if there's an error, and it's unrelated to the batch we sent, e.g.
+		// connection issue or something, we return here
+		if err != nil && !strings.Contains(err.Error(), unableToUpdateBlocklistErrStr) {
+			return
+		}
+
+		// otherwise if there's an error and the batchsize is larger than 1, we
+		// simply decrease the batch size and continue
+		if err != nil && batchSize > 1 {
+			bl.staticLogger.Tracef("Error occurred while blocking skylinks retrying with batch size %v, err: %v, retrying with smaller batch size...", batchSize, err)
+			batchSize /= blockBatchSizeDivisor
```

Comment:
nit: it would be nice to have a sanity check here that batchSize >= 1.

Reply:
this is solved now that we have the exit condition of the loop